gnome-initial-setup/gis-page-util.c

```diff
@@ -492,3 +491,94 @@ gis_page_util_show_demo_dialog (GisPage *page)
   gtk_window_set_modal (GTK_WINDOW (demo_dialog), TRUE);
   gtk_window_present (GTK_WINDOW (demo_dialog));
 }
+
+static gboolean
+run_dialog_cb (gpointer user_data)
+{
+  GtkDialog *dialog = GTK_DIALOG (user_data);
+  gtk_dialog_run (dialog);
+  return G_SOURCE_REMOVE;
+}
+
+static void
+on_reformatter_exited (GTask  *task,
+                       GError *error)
+{
+  GisPage *page = GIS_PAGE (g_task_get_source_object (task));
+  GisDriver *driver = page->driver;
+
+  gis_driver_show_window (driver);
+
+  if (error == NULL)
+    {
+      g_task_return_boolean (task, TRUE);
+    }
+  else
+    {
+      GtkWindow *toplevel = GTK_WINDOW (gtk_widget_get_toplevel (GTK_WIDGET (page)));
+      GtkWidget *message_dialog;
+
+      g_critical ("Error running the reformatter: %s", error->message);
+
+      message_dialog = gtk_message_dialog_new (toplevel,
+                                               GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT,
+                                               GTK_MESSAGE_ERROR,
+                                               GTK_BUTTONS_CLOSE,
+                                               _("Error running the reformatter: %s"), error->message);
+      g_idle_add_full (G_PRIORITY_DEFAULT_IDLE, run_dialog_cb,
+                       message_dialog, (GDestroyNotify) gtk_widget_destroy);
+
+      g_task_return_error (task, g_steal_pointer (&error));
+    }
+}
+
+static void
+reformatter_exited_cb (GObject      *source,
+                       GAsyncResult *result,
+                       gpointer      user_data)
+{
+  g_autoptr(GTask) task = G_TASK (user_data);
+  g_autoptr(GError) error = NULL;
+
+  g_subprocess_wait_check_finish (G_SUBPROCESS (source), result, &error);
+  on_reformatter_exited (task, g_steal_pointer (&error));
+}
+
+/**
+ * gis_page_util_run_reformatter:
+ *
+ * Launches the reformatter, and arranges for the assistant to be hidden for
+ * the duration of its runtime. @callback will be called when the reformatter
+ * exits.
+ *
+ * There is no corresponding _finish() function because (at present) neither
+ * caller actually cares about the result.
+ */
+void
+gis_page_util_run_reformatter (GisPage            *page,
+                               GAsyncReadyCallback callback,
+                               gpointer            user_data)
+{
+  g_autoptr(GTask) task = g_task_new (page, NULL, callback, user_data);
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) subprocess = NULL;
+  const gchar *locale = setlocale (LC_MESSAGES, NULL);
+  const gchar *command = "/usr/lib/eos-installer/gnome-image-installer";
+  g_autoptr(GError) error = NULL;
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_setenv (launcher, "LANG", locale, TRUE);
+  subprocess = g_subprocess_launcher_spawn (launcher, &error, command, NULL);
+
+  if (error)
+    {
+      on_reformatter_exited (task, g_steal_pointer (&error));
```

Comment:
Again you could early return here.

Reply:
I'm not sure this is easier to read with an early return, but ok.